### PR TITLE
Enable lints `library_private_types_in_public_api`, `sort_child_properties_last` and `use_key_in_widget_constructors`

### PR DESCRIPTION
**Pull request by asashour**
Fixes the warnings of the flutter lints [library_private_types_in_public_api](https://dart-lang.github.io/linter/lints/library_private_types_in_public_api.html), [sort_child_properties_last](https://dart-lang.github.io/linter/lints/sort_child_properties_last.html) and [use_key_in_widget_constructors](https://dart-lang.github.io/linter/lints/use_key_in_widget_constructors.html)

Fixes https://github.com/flutter/flutter/issues/102739

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [X] I signed the [CLA].
- [ ] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by stuartmorgan** [comment]
We run the analyzer on all of our code, include examples, on every run, and the tree is green, so there should not be any current issues. Are you proposing adding new lints?

If so--and if they are already in flutter/flutter--they need to be added to our analysis options. We don't change code to follow lints that we don't have enabled.

**Comment by asashour** [comment]
Thanks for the explanation.

The reason is because `flutter_lints` is not in `dev_dependencies` and there is no `analysis_options.yaml` in `image_picker` project.

For a user, who would create a new project, and copy the example, those lints would be reported, because the default flutter project contains `flutter_lints` and `analysis_options.yaml` referencing `flutter_lints`, which is the scenario that triggered the PR.

Should we proceed forward and add the those to the project? And should it be based on flutter `stable` or `master`?

Thanks.

**Comment by stuartmorgan** [comment]
> Should we proceed forward and add those to the project? And should it be based on flutter `stable` or `master`?

If you mean should `flutter_lints` be added to the repository, no. See https://github.com/flutter/flutter/issues/76229#issuecomment-1112319439

> I see now there is a `analysis_options.yaml` for all projects, which I guess should be modified.

Yes, this is where new lints should be added. I've checked, and
- use_key_in_widget_constructors
- library_private_types_in_public_api
- sort_child_properties_last

Are all in flutter/flutter's options, so we're just behind. I'm happy to review a PR that adds them.

prefer_const_constructors is already in our options, so I'm not clear on what you are fixing related to that.

**Comment by stuartmorgan** [comment]
> I don't know how to make `publishable` happy.

Every package you changed needs a CHANGELOG update; as you can see in the failure output, that's missing for many packages. If you hold off for a few days, I'm going to write a repository tool command to mass-update packages with identical CHANGELOG messages for cases like this, as it's something we clearly need for handling analysis updates.

I've also removed the "No version change:" entry from your PR description, as it was not correct. In cases where you are changing non-example code (which it looks like isn't many of them) and the CI flagged that you need a version change, that is correct. Please read https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version; the changes here don't meet any of those exemptions, and are changing the actual code that clients of the plugin will run.

**Comment by asashour** [comment]
Thanks for the feedback.

I read the [Contributing](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version) and have to say, I find it hard to understand this `NEXT`, and then `If you are adding a version change to a CHANGELOG that starts with NEXT, replace NEXT with the new version`.

Sure, take you rime, having a tool would be helpful for everyone.

**Comment by stuartmorgan** [comment]
> I read the [Contributing](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version) and have to say, I find it hard to understand this `NEXT`, and then `If you are adding a version change to a CHANGELOG that starts with NEXT, replace NEXT with the new version`.

Thanks for the feedback. I've added examples; hopefully it's clearer now.

**Comment by asashour** [comment]
The green tick is finally there :sweat_smile:!

There are two things I would like to hint about:
- It took me some time to figure out that there is a `tool_runner version-check` to verify it locally. Adding it to the wiki could be helpful.
- The local running of `version-check` gives a different result from the CI, possibly because it locally runs against the previous commits (of the same PR), while the server-one runs against the `head`. This could be also a potential enhancement, so that it always verifies against the `head` not matter which changes are locally committed.

**Comment by stuartmorgan** [comment]
> * The local running of `version-check` gives a different result from the CI, possibly because it locally runs against the previous commits (of the same PR), while the server-one runs against the `head`. 

You can pass `--base-sha` to control what it diffs against.

> This could be also a potential enhancement, so that it always verifies against the `head` not matter which changes are locally committed.

If it verified against `HEAD`, then it would only verify uncommitted changes, which would almost never be the desired behavior.

The default value is `merge-base` (as the help for that argument says), which is our best guess for the correct diff point, and usually what you want. If you've done something unusual locally that makes that the wrong diff point, there's no way for anyone but you to know what it should use instead.

File: packages/camera/camera_web/example/lib/main.dart [hunk 1]
```diff
@@ -2,12 +2,16 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+// ignore_for_file: public_member_api_docs
```
Comment (stuartmorgan) [inline packages/camera/camera_web/example/lib/main.dart:5]: This shouldn't be added; we want documentation in examples.

I know there are some files that have it, but that's a legacy issue that we shouldn't expand.

File: packages/camera/camera_windows/CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,8 @@
-## NEXT
+## 0.1.1
```
Comment (stuartmorgan) [inline packages/camera/camera_windows/CHANGELOG.md:1]: This should be 0.1.0+1

File: packages/camera/camera_windows/example/lib/main.dart [hunk 3]
```diff
@@ -2,18 +2,22 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+// ignore_for_file: public_member_api_docs
```
Comment (stuartmorgan) [inline packages/camera/camera_windows/example/lib/main.dart:5]: Same here, and throughout the rest of the PR.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Overall this looks great! Just a few minor things.

**Comment by asashour** [comment]
Thanks for your input.

> The default value is `merge-base` (as the help for that argument says), which is our best guess for the correct diff point, and usually what you want. If you've done something unusual locally that makes that the wrong diff point, there's no way for anyone but you to know what it should use instead.

I said something incorrect, sorry. I meant to be based on `main` latest commit, by comparing the working directory to it. So it doesn't matter if there are previous commits to the PR or not.

Interestingly, I have rebased the PR two days ago, and even now, with some local changes (before committing the version change of `camera_windows`), it seems that `merge-base` is 9 days ago, as shown in the below logs when the `version-check` is locally triggered.

```
[0:00] Running for camera/camera...
  Incorrectly updated version.
  HEAD: 0.9.4+22, d8f6ddbaad39e1001fe7fb3d7acd811ee66ea721: 0.9.4+20.
  Allowed versions: {1.0.0: NextVersionType.RELEASE, 0.10.0: NextVersionType.BREAKING_MAJOR, 0.9.5: NextVersionType.MINOR, 0.9.4+21: NextVersionType.PATCH}
```

All review points are resolved, thanks.

**Comment by stuartmorgan** [comment]
> I meant to be based on `main` latest commit, by comparing the working directory to it. So it doesn't matter if there are previous commits to the PR or not.

But it would matter when you updated main relative to when you rebased your PR. What you want is not the diff against main, but the diff against the last commit to main that is in your branch. Which is exactly what `merge-base` is for.

> it seems that `merge-base` is 9 days ago

Interesting. I looked into this and it looks like the tooling is passing an incorrect argument to `merge-base`. I've filed https://github.com/flutter/flutter/issues/103285

**Review by gaaclarke — COMMENTED** [review]
How do you want to do this now that we have owners?  `video_player` `shared_preferences` and `path_provider` all look good to me.

**Review by stuartmorgan — APPROVED** [review]
LGTM!

**Review by gaaclarke — APPROVED** [review]
lgtm

**Comment by Hixie** [comment]
test-exempt: has a test

**Comment by stuartmorgan** [comment]
Oops, sorry I landed that before you commented. I was going to land it, realized it needed an NTE, asked for it, context switched a couple of times, came back to this window, forgot there was a reason I hadn't landed it, and added the label :(

**Pull request merged**